File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/widgets/customization/BlazeReactWidgetLayoutMerger.kt
package com.blaze.flutterblazesdk.widgets.customization

import android.content.Context
import com.blaze.blazesdk.style.shared.models.BlazeObjectPositioning
import com.blaze.blazesdk.style.shared.models.BlazeObjectXPosition
import com.blaze.blazesdk.style.shared.models.BlazeObjectYPosition
import com.blaze.blazesdk.style.shared.models.blazeDp
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemBadgeStateStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemBadgeStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemImageAnimatedThumbnailStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemImageContainerBorderStateStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemImageContainerBorderStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemImageGradientOverlayStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemImageStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemStatusIndicatorStateStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemStatusIndicatorStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemTextStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetItemTitleStyle
import com.blaze.blazesdk.style.widgets.BlazeWidgetLayout
import com.blaze.flutterblazesdk.utils.parsing.gravityMergedWith
import com.blaze.flutterblazesdk.utils.parsing.mergedWith
import com.blaze.flutterblazesdk.utils.parsing.safeParseColor
import com.blaze.flutterblazesdk.utils.parsing.toFontResId
import com.blaze.flutterblazesdk.utils.parsing.toImageResId

fun BlazeWidgetLayout.mergedWith(
        customization: BlazeReactWidgetLayout?,
        context: Context
): BlazeWidgetLayout {
        customization ?: return this

        val merged = this

        merged.horizontalItemsSpacing =
                customization.horizontalItemsSpacing?.blazeDp ?: merged.horizontalItemsSpacing
        merged.verticalItemsSpacing =
                customization.verticalItemsSpacing?.blazeDp ?: merged.verticalItemsSpacing
        merged.itemRatio = customization.itemRatio ?: merged.itemRatio
        merged.columns = customization.columns ?: merged.columns
        merged.maxDisplayItemsCount =
                customization.maxDisplayItemsCount ?: merged.maxDisplayItemsCount
        merged.widgetItemStyle =
                this.widgetItemStyle.mergedWith(customization.widgetItemStyle, context)
        merged.margins = this.margins.mergedWith(customization.margins)
        return merged
}

fun BlazeWidgetItemStyle.mergedWith(
        customization: BlazeReactWidgetItemStyle?,
        context: Context
): BlazeWidgetItemStyle {
        customization ?: return this

        val merged = this

        merged.title = this.title.mergedWith(customization.title, context)
        merged.image = this.image.mergedWith(customization.image)
        merged.backgroundColor =
                safeParseColor(customization.backgroundColor) ?: this.backgroundColor
        merged.cornerRadius = customization.cornerRadius?.toInt()?.blazeDp ?: this.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?: this.cornerRadiusRatio
        merged.padding = this.padding.mergedWith(customization.margins)
        merged.statusIndicator =
                this.statusIndicator.mergedWith(customization.statusIndicator, context = context)
        merged.badge = this.badge.mergedWith(customization = customization.badge, context = context)

        return merged
}

fun BlazeWidgetItemStatusIndicatorStyle.mergedWith(
        customization: BlazeReactWidgetItemStatusIndicatorStyle?,
        context: Context
): BlazeWidgetItemStatusIndicatorStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: this.isVisible
        merged.position = this.position.mergedWith(customization.position)
        merged.margins = this.margins.mergedWith(customization.margins)
        merged.padding = this.padding.mergedWith(customization.statusTitlePadding)
        merged.liveUnreadState =
                this.liveUnreadState.mergedWith(customization.liveUnreadState, context = context)
        merged.liveReadState =
                this.liveReadState.mergedWith(customization.liveReadState, context = context)
        merged.unreadState =
                this.unreadState.mergedWith(customization.unreadState, context = context)
        merged.readState = this.readState.mergedWith(customization.readState, context = context)

        return merged
}

fun BlazeWidgetItemStatusIndicatorStateStyle.mergedWith(
        customization: BlazeReactWidgetItemStatusIndicatorStateStyle?,
        context: Context
): BlazeWidgetItemStatusIndicatorStateStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: merged.isVisible
        merged.backgroundColor =
                safeParseColor(customization.backgroundColor) ?: merged.backgroundColor
        merged.backgroundImageResId = customization.backgroundImage?.toImageResId(context)
        merged.textStyle = this.textStyle.mergedWith(customization.textStyle, context)
        merged.text = customization.text ?: merged.text
        merged.cornerRadius = customization.cornerRadius?.toInt()?.blazeDp ?: merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?: merged.cornerRadiusRatio
        merged.borderColor = safeParseColor(customization.borderColor) ?: merged.borderColor
        merged.borderWidth = customization.borderWidth?.blazeDp ?: merged.borderWidth

        return merged
}

fun BlazeWidgetItemTitleStyle.mergedWith(
        customization: BlazeReactWidgetItemTitleStyle?,
        context: Context
): BlazeWidgetItemTitleStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: this.isVisible
        merged.readState = this.readState.mergedWith(customization.readState, context)
        merged.unreadState = this.unreadState.mergedWith(customization.unreadState, context)
        merged.margins = this.margins.mergedWith(customization.margins)
        merged.position = this.position.mergedWith(customization.position)

        return merged
}

fun BlazeWidgetItemBadgeStyle.mergedWith(
        customization: BlazeReactWidgetItemBadgeStyle?,
        context: Context
): BlazeWidgetItemBadgeStyle {

        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: this.isVisible
        merged.position = this.position.mergedWith(customization.position)
        merged.margins = this.margins.mergedWith(customization.margins)
        merged.padding = this.padding.mergedWith(customization.titlePadding)
        merged.liveUnreadState =
                this.liveUnreadState.mergedWith(customization.liveUnreadState, context = context)
        merged.liveReadState =
                this.liveReadState.mergedWith(customization.liveReadState, context = context)
        merged.unreadState =
                this.unreadState.mergedWith(customization.unreadState, context = context)
        merged.readState = this.readState.mergedWith(customization.readState, context = context)

        return merged
}

fun BlazeWidgetItemBadgeStateStyle.mergedWith(
        customization: BlazeReactWidgetItemBadgeStateStyle?,
        context: Context
): BlazeWidgetItemBadgeStateStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: merged.isVisible
        merged.width = customization.width?.blazeDp ?: this.width
        merged.height = customization.height?.blazeDp ?: this.height
        merged.backgroundColor =
                safeParseColor(customization.backgroundColor) ?: merged.backgroundColor
        merged.backgroundImageResId = customization.backgroundImage?.toImageResId(context)
        merged.textStyle = this.textStyle.mergedWith(customization.textStyle, context)
        merged.text = customization.text ?: merged.text
        merged.cornerRadius = customization.cornerRadius?.toInt()?.blazeDp ?: merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?: merged.cornerRadiusRatio
        merged.borderColor = safeParseColor(customization.borderColor) ?: merged.borderColor
        merged.borderWidth = customization.borderWidth?.blazeDp ?: merged.borderWidth

        return merged
}

fun BlazeObjectPositioning.mergedWith(
        customization: BlazeReactObjectPositioning?
): BlazeObjectPositioning {
        customization ?: return this
        val merged = this

        merged.xPosition = this.xPosition.mergedWith(customization.xPosition)
        merged.yPosition = this.yPosition.mergedWith(customization.yPosition)
        return merged
}

fun BlazeObjectXPosition.mergedWith(
        customization: BlazeReactObjectXPosition?
): BlazeObjectXPosition {
        customization ?: return this

        return customization.mapToBlazeEnumClass()
}

fun BlazeObjectYPosition.mergedWith(
        customization: BlazeReactObjectYPosition?
): BlazeObjectYPosition {
        customization ?: return this

        return customization.mapToBlazeEnumClass()
}

fun BlazeWidgetItemTextStyle.mergedWith(
        customization: BlazeReactTitleStyle?,
        context: Context
): BlazeWidgetItemTextStyle {
        customization ?: return this

        val merged = this
        merged.fontResId = customization.font?.toFontResId(context)
        merged.textSize = customization.textSize ?: this.textSize
        merged.letterSpacing = customization.letterSpacing ?: this.letterSpacing
        merged.textColor = safeParseColor(customization.textColor) ?: this.textColor
        merged.lineHeight = customization.lineHeight?.blazeDp ?: this.lineHeight
        merged.maxLines = customization.maxLines ?: this.maxLines
        merged.gravity = this.gravity.gravityMergedWith(customization.textAlign)

        return merged
}

fun BlazeWidgetItemImageStyle.mergedWith(
        customization: BlazeReactWidgetItemImageStyle?
): BlazeWidgetItemImageStyle {
        customization ?: return this

        val merged = this

        merged.position = this.position.mergedWith(customization.position)
        merged.width = customization.width?.blazeDp ?: this.width
        merged.height = customization.height?.blazeDp ?: this.height
        merged.ratio = customization.ratio ?: this.ratio
        merged.cornerRadius = customization.cornerRadius?.blazeDp ?: this.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?: this.cornerRadiusRatio
        merged.border = this.border.mergedWith(customization.border)
        customization.thumbnailType?.let { merged.thumbnailType = it.mapToBlazeEnumClass() }
        merged.margins = this.margins.mergedWith(customization.margins)
        merged.animatedThumbnail =
                this.animatedThumbnail.mergedWith(customization.animatedThumbnail)
        merged.gradientOverlay = this.gradientOverlay.mergedWith(customization.gradientOverlay)

        return merged
}

fun BlazeWidgetItemImageGradientOverlayStyle.mergedWith(
        customization: BlazeReactWidgetItemImageGradientOverlayStyle?
): BlazeWidgetItemImageGradientOverlayStyle {
        customization ?: return this

        val merged = this
        merged.isVisible = customization.isVisible ?: merged.isVisible
        merged.position = merged.position.mergedWith(customization.position)

        // We switch between startColor and endColor to match the iOS implementation.
        // If this is aligned on future Android version we must switch them back!
        merged.startColor = safeParseColor(customization.endColor) ?: merged.startColor
        merged.endColor = safeParseColor(customization.startColor) ?: merged.endColor

        return merged
}

fun BlazeWidgetItemImageGradientOverlayStyle.BlazeGradientPosition.mergedWith(
        customization: BlazeReactWidgetItemImageGradientOverlayStyle.Position?
): BlazeWidgetItemImageGradientOverlayStyle.BlazeGradientPosition {
        customization ?: return this

        return customization.mapToBlazeEnumClass()
}

fun BlazeWidgetItemImageAnimatedThumbnailStyle.mergedWith(
        customization: BlazeReactWidgetItemImageAnimatedThumbnailStyle?
): BlazeWidgetItemImageAnimatedThumbnailStyle {
        customization ?: return this

        val merged = this
        merged.horizontalAnimationTriggerPercentage =
                customization.horizontalAnimationTriggerPercentage
                        ?: merged.horizontalAnimationTriggerPercentage
        merged.isEnabled = customization.isEnabled ?: merged.isEnabled

        return merged
}

fun BlazeWidgetItemImageStyle.BlazeImagePosition.mergedWith(
        customization: BlazeReactWidgetItemImagePosition?
): BlazeWidgetItemImageStyle.BlazeImagePosition {
        customization ?: return this

        return customization.mapToBlazeEnumClass()
}

fun BlazeWidgetItemImageContainerBorderStyle.mergedWith(
        customization: BlazeReactWidgetItemImageContainerBorderStyle?
): BlazeWidgetItemImageContainerBorderStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: merged.isVisible
        merged.liveReadState = merged.liveReadState.mergedWith(customization.liveReadState)
        merged.liveUnreadState = merged.liveUnreadState.mergedWith(customization.liveUnreadState)
        merged.readState = merged.readState.mergedWith(customization.readState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        return merged
}

fun BlazeWidgetItemImageContainerBorderStateStyle.mergedWith(
        customization: BlazeReactWidgetItemImageContainerBorderStateStyle?
): BlazeWidgetItemImageContainerBorderStateStyle {
        customization ?: return this

        val merged = this

        merged.isVisible = customization.isVisible ?: merged.isVisible
        merged.color = safeParseColor(customization.color) ?: merged.color
        merged.margin = customization.margin?.blazeDp ?: merged.margin
        merged.width = customization.width?.blazeDp ?: merged.width

        return merged
}

File: blaze_flutter_sdk/ios/Sources/Widgets/LayoutCustomization/WidgetLayout/BlazeReactWidgetLayoutMerger.swift
import Foundation
import BlazeSDK

extension BlazeWidgetLayout {
    func mergedWith(_ customization: BlazeReactWidgetLayout?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        merged.horizontalItemsSpacing = customization.horizontalItemsSpacing ?? merged.horizontalItemsSpacing
        merged.verticalItemsSpacing = customization.verticalItemsSpacing ?? merged.verticalItemsSpacing
        merged.itemRatio = customization.itemRatio ?? merged.itemRatio
        merged.columns = customization.columns ?? merged.columns
        merged.maxDisplayItemsCount = customization.maxDisplayItemsCount ?? merged.maxDisplayItemsCount
        merged.widgetItemStyle = merged.widgetItemStyle.mergedWith(customization.widgetItemStyle)
        merged.margins = merged.margins.mergedWith(customization.margins)
        return merged
    }
}

extension BlazeWidgetItemStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.title = merged.title.mergedWith(customization.title)
        merged.image = merged.image.mergedWith(customization.image)
        merged.backgroundColor = UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?? merged.cornerRadiusRatio
        merged.insets = merged.insets.mergedWith(customization.margins)
        merged.statusIndicator = merged.statusIndicator.mergedWith(customization.statusIndicator)
        merged.badge = merged.badge.mergedWith(customization.badge)
        merged.durationElement = merged.durationElement.mergedWith(customization.durationElement)
        
        return merged
    }
}

extension BlazeWidgetItemTitleStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemTitleStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.readState = merged.readState.mergedWith(customization.readState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        merged.position = merged.position.mergedWith(customization.position)
        merged.insets = merged.insets.mergedWith(customization.margins)
        return merged
    }
}

extension BlazeObjectPositioning {
    func mergedWith(_ customization: BlazeReactObjectPositioning?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.xPosition = merged.xPosition.mergedWith(customization.xPosition)
        merged.yPosition = merged.yPosition.mergedWith(customization.yPosition)

        return merged
    }
}

extension BlazeObjectXPosition {
    func mergedWith(_ customization: BlazeReactObjectXPosition?) -> Self {
        guard let customization else { return self }
        var results = BlazeSDK.BlazeObjectXPosition.centerX(offset: 0)
        
        switch customization {
        case .leadingToLeading:
            results = BlazeSDK.BlazeObjectXPosition.leadingToLeading(offset: 0)
        case .leadingToTrailing:
            results = BlazeSDK.BlazeObjectXPosition.leadingToTrailing(offset: 0)
        case .trailingToLeading:
            results = BlazeSDK.BlazeObjectXPosition.trailingToLeading(offset: 0)
        case .trailingToTrailing:
            results = BlazeSDK.BlazeObjectXPosition.trailingToTrailing(offset: 0)
        case .centerX:
            results = BlazeSDK.BlazeObjectXPosition.centerX(offset: 0)
        case .centerToLeading:
            results = BlazeSDK.BlazeObjectXPosition.centerToLeading(offset: 0)
        case .centerToTrailing:
            results = BlazeSDK.BlazeObjectXPosition.centerToTrailing(offset: 0)
        }
        
        return results
    }
}


extension BlazeObjectYPosition {
    func mergedWith(_ customization: BlazeReactObjectYPosition?) -> Self {
        guard let customization else { return self }
        var results = BlazeSDK.BlazeObjectYPosition.centerY(offset: 0)
        
        switch customization {
        case .topToBottom:
            results = BlazeSDK.BlazeObjectYPosition.topToBottom(offset: 0)
        case .bottomToBottom:
            results = BlazeSDK.BlazeObjectYPosition.topToBottom(offset: 0)
        case .bottomToTop:
            results = BlazeSDK.BlazeObjectYPosition.bottomToTop(offset: 0)
        case .centerToTop:
            results = BlazeSDK.BlazeObjectYPosition.centerToTop(offset: 0)
        case .centerY:
            results = BlazeSDK.BlazeObjectYPosition.centerY(offset: 0)
        case .centerToBottom:
            results = BlazeSDK.BlazeObjectYPosition.centerToBottom(offset: 0)
        case .topToTop:
            results = BlazeSDK.BlazeObjectYPosition.topToTop(offset: 0)
        }
        
        return results
    }
}

extension BlazeWidgetItemTextStyle {
    func mergedWith(_ customization: BlazeReactTitleStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.font = merged.font.fontWith(customization.font,
                                           textSize: customization.textSize)
        merged.letterSpacing = customization.letterSpacing ?? merged.letterSpacing
        merged.textColor = UIColor(hexString: customization.textColor) ?? merged.textColor
        merged.lineHeight = customization.lineHeight ?? merged.lineHeight
        merged.numberOfLines = customization.maxLines ?? merged.numberOfLines
        merged.alignment = merged.alignment.mergedWith(customization.textAlign)
        
        return merged
    }
}

extension BlazeWidgetItemImageStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemImageStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.position = merged.position.mergedWith(customization.position)
        merged.width = customization.width ?? merged.width
        merged.height = customization.height ?? merged.height
        merged.ratio = customization.ratio ?? merged.ratio
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?? merged.cornerRadiusRatio
        merged.border = merged.border.mergedWith(customization.border)
        merged.thumbnailType = merged.thumbnailType.mergedWith(customization.thumbnailType)
        merged.insets = merged.insets.mergedWith(customization.margins)
        merged.animatedThumbnail = merged.animatedThumbnail.mergedWith(customization.animatedThumbnail)
        merged.gradientOverlay = merged.gradientOverlay.mergedWith(customization.gradientOverlay)
        
        return merged
    }
}

extension BlazeSDK.BlazeWidgetItemImageGradientOverlayStyle {
    
    func mergedWith(_ customization: BlazeReactWidgetItemImageGradientOverlayStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.position = merged.position.mergedWith(customization.position)
        merged.startColor = UIColor(hexString: customization.startColor) ?? merged.startColor
        merged.endColor = UIColor(hexString: customization.endColor) ?? merged.endColor
        
        return merged
    }
}

extension BlazeSDK.BlazeWidgetItemImageGradientOverlayStyle.Position {
    func mergedWith(_ customization: BlazeReactWidgetItemImageGradientOverlayStyle.Position?) -> Self {
        guard let customization else { return self }
        
        let result: Self
        switch customization {
        case .top:
            result = .top
        case .center:
            result = .center
        case .bottom:
            result = .bottom
        }
        
        return result
    }
}

extension BlazeWidgetItemImageAnimatedThumbnailStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemAnimatedThumbnailStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.horizontalAnimationTriggerPercentage = customization.horizontalAnimationTriggerPercentage ?? merged.horizontalAnimationTriggerPercentage
        merged.isEnabled = customization.isEnabled ?? merged.isEnabled
        
        return merged
    }
}

extension BlazeWidgetItemImageStyle.Position {
    func mergedWith(_ customization: BlazeReactWidgetItemImagePosition?) -> Self {
        guard let customization else { return self }
        
        let result: Self
        switch customization {
        case .topStart:
            result = .topLeading
        case .topCenter:
            result = .topCenter
        case .topEnd:
            result = .topTrailing
        case .centerStart:
            result = .centerLeading
        case .center:
            result = .center
        case .centerEnd:
            result = .centerTrailing
        case .bottomStart:
            result = .bottomLeading
        case .bottomCenter:
            result = .bottomCenter
        case .bottomEnd:
            result = .bottomTrailing
        }
        
        return result
    }
}

extension BlazeWidgetItemImageContainerBorderStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemImageContainerBorderStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.liveReadState = merged.liveReadState.mergedWith(customization.liveReadState)
        merged.liveUnreadState = merged.liveUnreadState.mergedWith(customization.liveUnreadState)
        merged.readState = merged.readState.mergedWith(customization.readState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        
        return merged
    }
}

extension BlazeWidgetItemBadgeStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemBadgeStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.position = merged.position.mergedWith(customization.position)
        merged.insets = merged.insets.mergedWith(customization.margins)
        merged.liveUnreadState = merged.liveUnreadState.mergedWith(customization.liveUnreadState)
        merged.liveReadState = merged.liveReadState.mergedWith(customization.liveReadState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        merged.readState = merged.readState.mergedWith(customization.readState)
        
        return merged
    }
}

extension BlazeWidgetItemBadgeStateStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemBadgeStateStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.backgroundColor = UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.backgroundImage = UIImage.imageWith(customization.backgroundImage) ?? merged.backgroundImage
        merged.textStyle = merged.textStyle.mergedWith(customization.textStyle)
        merged.text = customization.text ?? merged.text
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?? merged.cornerRadiusRatio
        merged.borderColor = UIColor(hexString: customization.borderColor) ?? merged.borderColor
        merged.borderWidth = customization.borderWidth ?? merged.borderWidth
        merged.width = customization.width ?? merged.width
        merged.height = customization.height ?? merged.height
        
        return merged
    }
}

extension BlazeWidgetItemStatusIndicatorStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemStatusIndicatorStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.position = merged.position.mergedWith(customization.position)
        merged.insets = merged.insets.mergedWith(customization.margins)
        merged.statusTitlePadding = merged.statusTitlePadding.mergedWith(customization.statusTitlePadding)
        merged.liveUnreadState = merged.liveUnreadState.mergedWith(customization.liveUnreadState)
        merged.liveReadState = merged.liveReadState.mergedWith(customization.liveReadState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        merged.readState = merged.readState.mergedWith(customization.readState)
        
        return merged
    }
}

extension BlazeWidgetItemStatusIndicatorStateStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemStatusIndicatorStateStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.backgroundColor = UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.backgroundImage = UIImage.imageWith(customization.backgroundImage) ?? merged.backgroundImage
        merged.textStyle = merged.textStyle.mergedWith(customization.textStyle)
        merged.text = customization.text ?? merged.text
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?? merged.cornerRadiusRatio
        merged.borderColor = UIColor(hexString: customization.borderColor) ?? merged.borderColor
        merged.borderWidth = customization.borderWidth ?? merged.borderWidth
        
        return merged
    }
}

extension BlazeWidgetItemImageContainerBorderStateStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemImageContainerBorderStateStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.color = UIColor(hexString: customization.color) ?? merged.color
        merged.insets = customization.margin ?? merged.insets
        merged.width = customization.width ?? merged.width
        
        return merged
    }
}

extension BlazeWidgetItemImageStyle.ThumbnailType {
    func mergedWith(_ customization: BlazeReactThumbnailType?) -> Self {
        guard let customization else { return self }
        
        let result: BlazeWidgetItemImageStyle.ThumbnailType
        switch customization {
        case .squareIcon:
            result = .squareIcon
        case .verticalTwoByThree:
            result = .verticalTwoByThree
        case .custom:
            result = .custom
        }
        
        return result
    }
}

extension BlazeWidgetItemStyleOverrides {
    func mergedWith(_ customization: BlazeReactWidgetStylesOverrides?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.statusIndicator = merged.statusIndicator?.mergedWith(customization.statusIndicator)
        merged.imageBorder = merged.imageBorder?.mergedWith(customization.imageBorder)
        merged.badge = merged.badge?.mergedWith(customization.badge)
        
        return merged
    }
}

extension BlazeWidgetItemDurationElementStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemDurationElementStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.position = merged.position.mergedWith(customization.position)
        merged.insets = merged.insets.mergedWith(customization.insets)
        merged.contentPadding = merged.contentPadding.mergedWith(customization.contentPadding)
        merged.icon = UIImage.imageWith(customization.icon) ?? merged.icon
        merged.liveUnreadState = merged.liveUnreadState.mergedWith(customization.liveUnreadState)
        merged.liveReadState = merged.liveReadState.mergedWith(customization.liveReadState)
        merged.unreadState = merged.unreadState.mergedWith(customization.unreadState)
        merged.readState = merged.readState.mergedWith(customization.readState)
        
        return merged
    }
}

extension BlazeWidgetItemDurationElementStateStyle {
    func mergedWith(_ customization: BlazeReactWidgetItemDurationElementStyle.BlazeReactWidgetItemDurationElementStateStyle?) -> Self {
        guard let customization else { return self }
        
        var merged = self
        
        merged.isVisible = customization.isVisible ?? merged.isVisible
        merged.backgroundColor = UIColor(hexString: customization.backgroundColor) ?? merged.backgroundColor
        merged.backgroundImage = UIImage.imageWith(customization.backgroundImage) ?? merged.backgroundImage
        merged.textStyle = merged.textStyle.mergedWith(customization.textStyle)
        merged.cornerRadius = customization.cornerRadius ?? merged.cornerRadius
        merged.cornerRadiusRatio = customization.cornerRadiusRatio ?? merged.cornerRadiusRatio
        merged.borderColor = UIColor(hexString: customization.borderColor) ?? merged.borderColor
        merged.borderWidth = customization.borderWidth ?? merged.borderWidth
        
        return merged
    }
} 